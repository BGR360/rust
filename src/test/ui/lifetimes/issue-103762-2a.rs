// #103762: cases that pre-1.64 rejected that we should keep rejecting.

#![allow(bare_trait_objects)]

trait Trait {}

// Case 1: '_ in trait object bound without any other anonymous/elided lifetimes.
// Pre-1.64 rejected this, but 1.64 did not. Make sure it does.
fn a1(_: Box<dyn Trait + '_>) -> &str { loop {} } //~ ERROR missing lifetime specifier
fn a2(_: Box<Trait + '_>) -> &str { loop {} } //~ ERROR missing lifetime specifier

// Case 2: Inside generic bound.
// Pre-1.64 and 1.64 rejected this, should still reject.
fn b1<T: Trait + '_>(_: &T) -> &str { loop {} } //~ ERROR `'_` cannot be used here
fn b2<T>(_: &T) -> &str where T: Trait + '_ { loop {} } //~ ERROR `'_` cannot be used here

// Case 3.a: Inside impl Trait without any other anonymous/elided lifetimes.
//  * Pre-1.64 rejected this because it thought `+ '_` needed to be a named lifetime,
//    and did not report on `&str`.
//  * 1.64 rejected this because of unstable anonymous_lifetime_in_impl_trait,
//    and because `&str` needs a named lifetime.
// It makes sense to keep doing what 1.64 did.
fn c1(_: impl Trait + '_) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} } //~ ERROR missing lifetime specifier

// Case 3.b: Inside impl Trait with other anonymous/elided lifetimes.
//  * Pre-1.64 rejected these because it thought `+ '_` needed to be a named lifetime.
//  * 1.64 rejected these because of unstable anonymous_lifetime_in_impl_trait,
//    and because it thought `&str` needed a named lifetime parameter.
// It should continue reporting on the unstable feature, but not on `&str`, since we
// shouldn't count `+ '_` as a new elision candidate.
fn c2(_: &(impl Trait + '_)) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} }
fn c3(_: &i32, _: impl Trait + '_) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} }
fn c4(_: impl Trait + '_, _: &i32) //~ ERROR anonymous lifetimes in `impl Trait` are unstable
    -> &str { loop {} }

fn main() {}
