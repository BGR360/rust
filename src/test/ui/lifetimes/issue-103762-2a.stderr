error[E0106]: missing lifetime specifier
  --> $DIR/issue-103762-2a.rs:9:34
   |
LL | fn a1(_: Box<dyn Trait + '_>) -> &str { loop {} }
   |                                  ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime
   |
LL | fn a1(_: Box<dyn Trait + '_>) -> &'static str { loop {} }
   |                                   +++++++

error[E0106]: missing lifetime specifier
  --> $DIR/issue-103762-2a.rs:10:30
   |
LL | fn a2(_: Box<Trait + '_>) -> &str { loop {} }
   |                              ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime
   |
LL | fn a2(_: Box<Trait + '_>) -> &'static str { loop {} }
   |                               +++++++

error[E0637]: `'_` cannot be used here
  --> $DIR/issue-103762-2a.rs:14:18
   |
LL | fn b1<T: Trait + '_>(_: &T) -> &str { loop {} }
   |                  ^^ `'_` is a reserved lifetime name

error[E0637]: `'_` cannot be used here
  --> $DIR/issue-103762-2a.rs:15:42
   |
LL | fn b2<T>(_: &T) -> &str where T: Trait + '_ { loop {} }
   |                                          ^^ `'_` is a reserved lifetime name

error[E0106]: missing lifetime specifier
  --> $DIR/issue-103762-2a.rs:24:8
   |
LL |     -> &str { loop {} }
   |        ^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'static` lifetime
   |
LL |     -> &'static str { loop {} }
   |         +++++++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2a.rs:23:23
   |
LL | fn c1(_: impl Trait + '_)
   |                       ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn c1<'a>(_: impl Trait + '_'a )
   |      ++++                   ++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2a.rs:32:25
   |
LL | fn c2(_: &(impl Trait + '_))
   |                         ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn c2<'a>(_: &(impl Trait + '_'a ))
   |      ++++                     ++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2a.rs:34:32
   |
LL | fn c3(_: &i32, _: impl Trait + '_)
   |                                ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn c3<'a>(_: &i32, _: impl Trait + '_'a )
   |      ++++                            ++

error[E0658]: anonymous lifetimes in `impl Trait` are unstable
  --> $DIR/issue-103762-2a.rs:36:23
   |
LL | fn c4(_: impl Trait + '_, _: &i32)
   |                       ^^ expected named lifetime parameter
   |
   = help: add `#![feature(anonymous_lifetime_in_impl_trait)]` to the crate attributes to enable
help: consider introducing a named lifetime parameter
   |
LL | fn c4<'a>(_: impl Trait + '_'a , _: &i32)
   |      ++++                   ++

error: aborting due to 9 previous errors

Some errors have detailed explanations: E0106, E0637, E0658.
For more information about an error, try `rustc --explain E0106`.
